### PR TITLE
Offer Docsy as a single Hugo Module

### LINKED ISSUE #1120: Offer Docsy as a single Hugo Module
The intent would be to **streamline use of Docsy for Hugo Module users**, something which (IMHO) makes sense to do now that [Git submodules]() were removed in [0.4.0](https://github.com/google/docsy/blob/main/CHANGELOG.md#040), and NPM-package support is available (#949).

@deining - I can't find the thread, but at some point we discussed the possibility of eliminating the `dependencies` module, which was introduced as a means of addressing:

- #848

This issue is to record that intent, and offer a place for discussing it.

### Tasks

- [x] Module rework: #1217
  - [x] #173 
- [x] #1730
- [ ] #1731
- [ ] User guide updates - #1743
- [ ] Changelog entry marking this as a breaking change for Hugo-module users of Docsy - #1743

### References

- https://github.com/golang/go/issues/37397

### PR DESCRIPTION
**Pull request by deining**
This PR merges the module `docsy` and `docsy/dependencies` into one single module in order to streamline use of Docsy for Hugo Module users. This is a first  take at #1120.

### CODE REVIEW

**Comment by deining** [comment]
Netlify preview fails, but this was expected.

**Comment by chalin** [comment]
@deining - thanks for this. I'm willing to give this a spin. Could you please resolve conflicts (though maybe it is best to wait for #1218 to be merged first).

FYI, v0.5.0 will be released soon. This PR is targeted for a subsequent release (probably v0.6.0).

**Comment by deining** [comment]
> Could you please resolve conflicts.

Done.

> FYI, v0.5.0 will be released soon. This PR is targeted for a subsequent release (probably v0.6.0).

Great. I can contribute to the documentation and will assist in case of any questions/issues.

**Comment by deining** [comment]
Netlify preview (git submodule apporach) is fine now, so this PR is ready for review and subsequent merging now. @chalin: awaiting the results of your review!

**Comment by chalin** [comment]
Thanks! Will get to this as soon as I can.

**Review by chalin — COMMENTED** [review]
@deining - Shouldn't `github.com/google/docsy/dependencies` entries be removed from `go.mod` and `go.sum`?

Also running `hugo mod tidy` would be a good idea as well, no?

File: package.json [hunk 1]
```diff
@@ -18,6 +18,7 @@
     "docs-install": "npm run _cd:docs -- npm install",
     "fix:format": "npm run _check:format -- --write",
     "get:submodule": "set -x && git submodule update --init ${DEPTH:- --depth 1}",
+    "postinstall": "cd .. && npx mkdirp github.com/FortAwesome/Font-Awesome && npx mkdirp github.com/twbs/bootstrap",
```
Comment (chalin) [inline package.json:21]:
```suggestion
    "postinstall": "npx -no mkdirp ../github.com/FortAwesome/Font-Awesome ../github.com/twbs/bootstrap",
```

Comment (deining) [reply package.json:21]: Do you mean npx `--no` (two dashes)?

Comment (chalin) [reply package.json:21]: Oops, yes. Fixed.

Comment (deining) [reply package.json]: Resolved with commit 396d76d5358019fa7bb4a5146bb7811ef7804401 (as suggested).

**Comment by deining** [comment]
> @deining - Shouldn't `github.com/google/docsy/dependencies` entries be removed from `go.mod` and `go.sum`?

Yes, very true.

> Also running `hugo mod tidy` would be a good idea as well, no?

Done with commit [396d76d](https://github.com/google/docsy/pull/1217/commits/396d76d5358019fa7bb4a5146bb7811ef7804401).

**Comment by chalin** [comment]
Thanks for the updates.

Have you tested this locally, and if so how? I can't get it to work, whether with using a (slightly modified) version of `docsy-example` or starting from scratch, again following a modified version of the instructions https://www.docsy.dev/docs/get-started/docsy-as-module/start-from-scratch/.

**Comment by deining** [comment]
> Thanks for the updates.

You are welcome.

> Have you tested this locally,

Yes.


> and if so how?

I started a new site from scratch, created a clone of docsy beside the newly created site repo and used `hugo.work` to refer to the local docsy theme repo. Similar do what I proposed in another [PR](https://github.com/google/docsy-example/pull/194) from me. Everything ran smoothly, the site was up and running in no time 😄 .

> I can't get it to work

I prepare a branch on my fork of the docsy-example repo. Please try this:

```
git clone -b docsy-as-single-module https://github.com/deining/docsy-example docsy-as-single-module-test
hugo server
```

This works for me out of the box. Can you please give this a try?

File: assets/bootstrap/scss/_vendor/_rfs.scss
Comment (chalin) [inline assets/bootstrap/scss/_vendor/_rfs.scss:1]: Can we not simply place `_rfs.scss` under `assets/vendor/bootstrap/scss/vendor` and avoid corresponding mount point in `hugo.toml`?

(And move the README.* files appropriately too?)

**Comment by chalin** [comment]
> This works for me out of the box. Can you please give this a try?

Yes, it works, thanks!

I also got it to work with a `docsy.work` file.

File: assets/bootstrap/scss/_vendor/_rfs.scss
Comment (deining) [reply assets/bootstrap/scss/_vendor/_rfs.scss:1]: > Can we not simply place `_rfs.scss` under `assets/vendor/bootstrap/scss/vendor` and avoid corresponding mount point in `hugo.toml`?
> 
> (And move the README.* files appropriately too?)

No, we can't. As stated in `README.txt`:

```
This folder was added in order to work around a known bug in Go’s module management.
For details, please refer to: https://github.com/golang/go/issues/37397
```

Folders named `vendor` are disregarded by Go's module management. The only way to bring a folder with this name into the projetc's directory structure is via mouting. I would love to simplify that, too, but I don't see any other way.

**Comment by deining** [comment]
> > This works for me out of the box. Can you please give this a try?
> 
> Yes, it works, thanks!
> 
> I also got it to work with a `docsy.work` file.

Great to hear that!

Do you deem this PR production ready and ready for subsequent merging?
What is still missing and what are the next steps?

File: assets/bootstrap/scss/_vendor/_rfs.scss
Comment (chalin) [reply assets/bootstrap/scss/_vendor/_rfs.scss:1]: It is my understanding that the problem with `vendor` directories doesn't apply to top-level directories named "vendor".

If I make the change I described earlier, then `docsy-example` built with my locally modified docsy works just fine AFAICT.

WDYT? Can you try?

**Comment by chalin** [comment]
> Do you deem this PR production ready and ready for subsequent merging?
> What is still missing and what are the next steps?

Almost there! I'd like to address https://github.com/google/docsy/pull/1217#discussion_r1389230896 first.

File: assets/bootstrap/scss/_vendor/_rfs.scss
Comment (deining) [reply assets/bootstrap/scss/_vendor/_rfs.scss:1]: > It is my understanding that the problem with `vendor` directories doesn't apply to top-level directories named "vendor".
> 
> If I make the change I described earlier, then `docsy-example` built with my locally modified docsy works just fine AFAICT.
> 
> WDYT? Can you try?

Yes indeed, we most likely can avoid the mount point indeed and simplify things here. At the time when I implemented docsy as hugo module system we still had git sub repos in place, that's the reason for the mount point. But that has changed meanwhile.

I will try during weekend.

Comment (deining) [reply assets/bootstrap/scss/_vendor/_rfs.scss:1]: > > Can we not simply place `_rfs.scss` under `assets/vendor/bootstrap/scss/vendor` and avoid corresponding mount point in `hugo.toml`?

I tested this, no we cannot.
Yes. it works as local copy, that's what you saw. But as soon as you pull in the module from remote, it breaks. And this is the reason: in the hugo module cache, have a look at the local theme clone: there is folder `assets/`  but no folder `assets/vendor/`. This breaks the build. I don't see any way to work around this. To make it short, I strongly advocate to leave the PR as is. I don't see any way to make it work, and I'm afraid we may run into regressions when altering the code here.

Comment (chalin) [reply assets/bootstrap/scss/_vendor/_rfs.scss:1]: Thanks for double checking. I agree, let's get this in as is. (I have some ideas, but they can wait.)

How can I check what's in hugo module cache?

**Review by chalin — APPROVED** [review]
👍🏻

**Comment by chalin** [comment]
In case you missed my nested comment @deining:

> How can I check what's in hugo module cache?

**Pull request merged**